Sources/TuistKit/Generator/WorkspaceGenerator.swift

```diff
@@ -52,47 +52,61 @@ final class WorkspaceGenerator: WorkspaceGenerating {
+        let xcworkspace = XCWorkspace(data: workspaceData)
 
-        /// Manifests
+        func recursiveChildElement(element: Workspace.Element) throws -> XCWorkspaceDataElement {
+            switch element {
+            case let .file(path: filePath):
+                return workspaceFileElement(path: filePath.relative(to: path))
 
-        let manifestFiles: [XCWorkspaceDataElement] = [Manifest.workspace, Manifest.setup]
-            .lazy
-            .map(pipe(get(\.fileName), path.appending))
-            .filter(fileHandler.exists)
-            .map { $0.relative(to: path) }
-            .map(workspaceFileElement)
+            case let .group(name: name, contents: contents):
 
-        workspace.data.children.append(contentsOf: manifestFiles)
+                let location = XCWorkspaceDataElementLocationType.container("")
+                let childData = XCWorkspaceDataGroup(
+                    location: location,
+                    name: name,
+                    children: try contents.map(recursiveChildElement)
+                )
 
-        /// Projects
-        var workspaceElements = [XCWorkspaceDataElement]()
+                return .group(childData)
 
-        try graph.projects.forEach { project in
-            let sourceRootPath = try projectDirectoryHelper.setupProjectDirectory(project: project,
-                                                                                  directory: directory)
-            let generatedProject = try projectGenerator.generate(project: project,
-                                                                 options: options,
-                                                                 graph: graph,
-                                                                 sourceRootPath: sourceRootPath)
+            case let .project(path: projectPath):
 
-            let relativePath = generatedProject.path.relative(to: path)
-            workspaceElements.append(workspaceFileElement(path: relativePath))
+                let project = graph.projects[projectPath]!
+
+                let sourceRootPath = try projectDirectoryHelper.setupProjectDirectory(
+                    project: project,
+                    directory: directory
+                )
+
+                let generatedProject = try projectGenerator.generate(
+                    project: project,
+                    options: options,
+                    graph: graph,
+                    sourceRootPath: sourceRootPath
+                )
+
+                let relativePath = generatedProject.path.relative(to: path)
+
+                return workspaceFileElement(path: relativePath)
+            }
         }
-        workspaceData.children.append(contentsOf: workspaceElements.sorted(by: workspaceDataElementSort))
 
-        try write(xcworkspace: workspace, to: workspacePath)
+        xcworkspace.data.children.append(contentsOf: try workspace.contents.map(recursiveChildElement))
```

Comment:
Am I wrong assuming that this could result in duplicated projects in the workspace if the project has already been defined in the `Workspace.swift`?

Reply:
AFAIK that will only happen if the project is actually referenced twice inside of the Workspace.swift file. With the changes I want to make it won't be possible to do so.

Reply:
Even in the following scenario?

```
Workspace
  ~> Project 1
  ~> Project 2
      ~> Project 1 (dependency)
```
Would Project 1 be duplicated?